bokehjs/src/lib/models/glyphs/multi_polygons.ts

```diff
@@ -30,48 +28,93 @@ export class MultiPolygonsView extends GlyphView {
   model: MultiPolygons
   visuals: MultiPolygons.Visuals
 
-  protected _index_data(): SpatialIndex {
-    const points = []
-    for (let i = 0, end = this._xs.length; i < end; i++) {
-      for (let j = 0, endj = this._xs[i].length; j < endj; j++) {
-        const xs = this._xs[i][j][0]  // do not use holes
-        const ys = this._ys[i][j][0]  // do not use holes
+  protected _hole_index: SpatialIndex
+
+  protected _index_data(index: SpatialIndex): void {
```

Comment:
Both `MultiPolygons` and `Patches` produced index that was larger than data arrays, resulting in potential out of bounds access and duplicated work (in particular duplicated rendering). I changed this to make sure all glyphs produce index consistently. However, I see that there may be a utility of having a more detailed index or a multi-level index, but that will have to be implemented separately to make sure that we don't break consistency and produce correct results.